Plugin resource reinstalls plugins on every Puppet run.

## Information
* Module version: 6.2.2
* Puppet version: 4.10.10
* OS and version: Redhat Enterprise Linux 6.9

## Bug description

Plugin resource reinstalls plugin(s) on every Puppet run. 

$elasticsearch_version = '6.2.4'

elasticsearch::plugin { "elasticsearch/x-pack/${elasticsearch_version}":
  url => "http://puppetrepo/elasticsearch/x-pack-${elasticsearch_version}.zip",
  instances => 'production',
}

[puppetrun.txt](https://github.com/elastic/puppet-elasticsearch/files/1995534/puppetrun.txt)
[puppetrun-debug.txt](https://github.com/elastic/puppet-elasticsearch/files/1995535/puppetrun-debug.txt)

I am seeing something that might be related, as I'm trying to install xpack for 6.2.2, but my puppet manifest is slightly different....as i am following the way this are setup in the README.....
  class { 'elasticsearch':
    manage_repo   => true,
    repo_version  => '6.x',
    version       => $esversion,
    config        => $config_hash,
    init_defaults => $init_hash,
    require       => Class['java'],
    security_plugin => 'x-pack',
  }
  elasticsearch::plugin { 'x-pack': instances => 'es-01' }

This is actually causing other plugins to not install properly and is breaking the installation.
`
Exception in thread "main" java.lang.IllegalStateException: Could not load plugin descriptor for existing plugin [.name]. Was the plugin built before 2.0?
	at org.elasticsearch.plugins.PluginsService.getPluginBundles(PluginsService.java:352)
	at org.elasticsearch.plugins.PluginsService.getPluginBundles(PluginsService.java:328)
	at org.elasticsearch.plugins.InstallPluginCommand.jarHellCheck(InstallPluginCommand.java:600)
	at org.elasticsearch.plugins.InstallPluginCommand.loadPluginInfo(InstallPluginCommand.java:588)
	at org.elasticsearch.plugins.InstallPluginCommand.installPlugin(InstallPluginCommand.java:707)
	at org.elasticsearch.plugins.InstallPluginCommand.install(InstallPluginCommand.java:623)
	at org.elasticsearch.plugins.InstallPluginCommand.execute(InstallPluginCommand.java:223)
	at org.elasticsearch.plugins.InstallPluginCommand.execute(InstallPluginCommand.java:212)
	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86)
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
	at org.elasticsearch.cli.MultiCommand.execute(MultiCommand.java:75)
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
	at org.elasticsearch.cli.Command.main(Command.java:90)
	at org.elasticsearch.plugins.PluginCli.main(PluginCli.java:48)
Caused by: java.nio.file.FileSystemException: /usr/share/elasticsearch/plugins/x-pack/.name/plugin-descriptor.properties: Not a directory
`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@
 * Service files for removed instances were previously set to ensure => absent on removal. Because this limits Puppet's ability to verify that the named service is running or not, these service files are always present now whether an instance is set to present or absent.
 * The service defaults file now enforces user/group ownership inline with the service user runtime.
 * The `scripts` configuration directory is now recursively copied instead of symlinked to avoid Elasticsearch security manager permission errors.
+* X-Pack and other meta-plugins are now properly detected as installed by the native plugin provider.
 
 ## 6.2.2 (March 13, 2018)
 

--- a/lib/puppet/provider/elastic_plugin.rb
+++ b/lib/puppet/provider/elastic_plugin.rb
@@ -25,8 +25,8 @@ class Puppet::Provider::ElasticPlugin < Puppet::Provider
     # plugin descriptor file, which each plugin should have. We must wildcard
     # the name to match meta plugins, see upstream issue for this change:
     # https://github.com/elastic/elasticsearch/pull/28022
-    properties = Dir[File.join(@resource[:plugin_dir], plugin_path, '*plugin-descriptor.properties')]
-    return false if properties.empty?
+    properties_files = Dir[File.join(@resource[:plugin_dir], plugin_path, '**', '*plugin-descriptor.properties')]
+    return false if properties_files.empty?
 
     begin
       # Use the basic name format that the plugin tool supports in order to
@@ -36,24 +36,31 @@ class Puppet::Provider::ElasticPlugin < Puppet::Provider
       # Naively parse the Java .properties file to check version equality.
       # Because we don't have the luxury of installing arbitrary gems, perform
       # simple parse with a degree of safety checking in the call chain
-      installed_version = IO.readlines(properties.first).map(&:strip).reject do |line|
-        line.start_with?('#') or line.empty?
-      end.map do |property|
-        property.split('=')
-      end.reject do |pairs|
-        pairs.length != 2
-      end.to_h['version']
+      #
+      # Note that x-pack installs "meta" plugins which bundle multiple plugins
+      # in one. Therefore, we need to find the first "sub" plugin that
+      # indicates which version of x-pack this is.
+      properties = properties_files.sort.map do |prop_file|
+        IO.readlines(prop_file).map(&:strip).reject do |line|
+          line.start_with?('#') or line.empty?
+        end.map do |property|
+          property.split('=')
+        end.reject do |pairs|
+          pairs.length != 2
+        end.to_h
+      end.find { |prop| prop.key? 'version' }
 
-      if installed_version != plugin_version
+      if properties and properties['version'] != plugin_version
         debug "Elasticsearch plugin #{@resource[:name]} not version #{plugin_version}, reinstalling"
         destroy
         return false
       end
     rescue ElasticPluginParseFailure
-      # If there is no version string, we do not check version equality
-      debug "No version found in #{@resource[:name]}, not enforcing any version"
+      debug "Failed to parse plugin version for #{@resource[:name]}"
     end
 
+    # If there is no version string, we do not check version equality
+    debug "No version found in #{@resource[:name]}, not enforcing any version"
     true
   end
 

